lib/pdf/core/embedded_files.rb

```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+module PDF
+  module Core
+    module EmbeddedFiles #:nodoc:
+      # The maximum number of children to fit into a single node in the
+      # EmbeddedFiles tree.
+      NAME_TREE_CHILDREN_LIMIT = 20 #:nodoc:
+
+      # The EmbeddedFiles name tree in the Name dictionary (see
+      # Prawn::Document::Internal#names). This name tree is used to store named
+      # embedded files (PDF spec 3.10.3). (For more on name trees, see section
+      # 3.8.4 in the PDF spec.)
+      #
+      def embedded_files
+        names.data[:EmbeddedFiles] ||= ref!(
+          PDF::Core::NameTree::Node.new(self, NAME_TREE_CHILDREN_LIMIT)
+        )
+      end
+
+      # Adds a new embedded file to the EmbeddedFiles name tree
+      # (see #embedded_files). The +reference+ parameter will be converted into
+      # a PDF::Core::Reference if it is not already one.
+      #
+      def add_embedded_file(name, reference)
+        reference = ref!(reference) unless reference.is_a?(PDF::Core::Reference)
```

Comment:
IIRC, `ref!` woudl return the same reference if the arg is already a reference so there's no need for the guard.

Reply:
I took a look into this and the guard is needed. Probably you are talking about the `push` method of the store in `ref`. So `ref!` always creates a reference.

Reply:
I mean, yes I was talking about `ObjectStore#push`. `ref!` calls `ObjectStore#ref` and that calls `ObjectStore#push` and it makes sure the argument is a reference. Unless I'm missing something that is the only way it can happen. But I wonder what I might be missing since you've checked and apparently found something different.

Reply:
From what I see on the code, `ObjectStore#push` has the parameter `*args` and checks for `args.first.is_a?(PDF::Core::Reference)`, returning the reference in that case. The problem is that `ObjectStore#ref` always passes an integer as the first parameter, the `identifier`. Maybe we could also check the second parameter and satisfy that behaviour.

Reply:
You're right. Apologies.

That API should probably be revised but that is clearly out of scope for this work.